thunderbirdparser/src/org/sleuthkit/autopsy/thunderbirdparser/EmailMessageThreader.java

```diff
@@ -0,0 +1,674 @@
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Given a list of email messages arranges the message into threads using the
+ * message reference lists.
+ *
+ * This threader is based heavely off of the algorithum found at
+ * <a href="https://www.jwz.org/doc/threading.html">
+ * "message threading." by Jamie Zawinski</a>
+ *
+ */
+final class EmailMessageThreader {
+
+    private int bogus_id_count = 0;
+
+    public void threadMessages(List<EmailMessage> emailMessages, String threadIDPrefix) {
+        HashMap<String, Container> id_table = createIDTable(emailMessages);
+        Set<Container> rootSet = getRootSet(id_table);
+
+        pruneEmptyContainers(rootSet);
+
+        Set<Container> finalSet = groupBySubject(rootSet);
+
+        assignThreadIDs(finalSet, threadIDPrefix);
+    }
+
+    /**
+     * Walks the list of emailMessages creating a Container object for each
+     * unique message ID found. Adds the emailMessage to the container where
+     * possible.
+     *
+     * @param emailMessages
+     *
+     * @return - HashMap of all message where the key is the message-ID of the message
+     */
+    private HashMap<String, Container> createIDTable(List<EmailMessage> emailMessages) {
+        HashMap<String, Container> id_table = new HashMap<>();
+
+        for (EmailMessage message : emailMessages) {
+            String messageID = message.getMessageID();
+
+            // Check the id_table for an existing Container for message-id
+            Container container = id_table.get(messageID);
+
+            // An existing container for message-id was found
+            if (container != null) {
+                // If the existing Container has a message already assocated with it
+                // emailMessage maybe a duplicate, so we don't lose the existance of 
+                // the duplicate message assign it a bogus message-id
+                if (container.hasMessage()) {
+                    messageID = String.format("<Bogus-id: %d >", bogus_id_count++);
+                    container = null;
+                } else {
+                    container.setMessage(message);
+                }
+            }
+
+            if (container == null) {
+                container = new Container(message);
+                id_table.put(messageID, container);
+            }
+
+            processMessageReferences(message, container, id_table);
+        }
+
+        return id_table;
+    }
+
+    /**
+     * Loops throught message's list of references, creating objects as needed
+     * and setting up the parent child relationships amoung the messages.
+     *
+     * @param message   The current email messags
+     * @param container Container object for message
+     * @param id_table  Hashtable of known message-id\container pairs
+     */
+    void processMessageReferences(EmailMessage message, Container container, Map<String, Container> id_table) {
+        List<String> referenceList = message.getReferences();
```

Comment:
The API docs for getReferences() seem to be incorrect (before this PR). They say that we return empty string.  For consistency, I think getReferences should return an empty list if there are no references instead of returning null.

Reply:
Let me clarify, its kind of assumed to be in parent order and it might be.  If we wanted to get really fancy with the algorithm the threads should be sorted by date at the very end.    I think saying that this list could be in a particular order is misleading.

The final step of the algorithm which is optional, is to sort the threads by date to put them in true order.  The documents suggest that it could also be done by the caller\client of this code.

I will document that the threads will not be in the correct time order.